src/main/java/rlbotexample/output/ControlsOutput.java

```diff
@@ -30,76 +33,97 @@
     private boolean slideDepressed;
     private boolean useItemDepressed;
 
-    public ControlsOutput() {
-    }
-
-    public ControlsOutput withSteer(float steer) {
-        this.steer = clamp(steer);
-        return this;
-    }
-
-    public ControlsOutput withPitch(float pitch) {
-        this.pitch = clamp(pitch);
-        return this;
-    }
-
-    public ControlsOutput withYaw(float yaw) {
-        this.yaw = clamp(yaw);
-        return this;
-    }
-
-    public ControlsOutput withRoll(float roll) {
-        this.roll = clamp(roll);
-        return this;
-    }
-
-    public ControlsOutput withThrottle(float throttle) {
-        this.throttle = clamp(throttle);
-        return this;
-    }
-
-    public ControlsOutput withJump(boolean jumpDepressed) {
-        this.jumpDepressed = jumpDepressed;
-        return this;
-    }
-
-    public ControlsOutput withBoost(boolean boostDepressed) {
-        this.boostDepressed = boostDepressed;
-        return this;
-    }
-
-    public ControlsOutput withSlide(boolean slideDepressed) {
-        this.slideDepressed = slideDepressed;
-        return this;
-    }
-
-    public ControlsOutput withUseItem(boolean useItemDepressed) {
-        this.useItemDepressed = useItemDepressed;
-        return this;
-    }
-
-    public ControlsOutput withJump() {
-        this.jumpDepressed = true;
-        return this;
-    }
-
-    public ControlsOutput withBoost() {
-        this.boostDepressed = true;
-        return this;
-    }
-
-    public ControlsOutput withSlide() {
-        this.slideDepressed = true;
-        return this;
-    }
-
-    public ControlsOutput withUseItem() {
-        this.useItemDepressed = true;
-        return this;
-    }
-
-    private float clamp(float value) {
-        return Math.max(-1, Math.min(1, value));
+    public ControlsOutput() { }
+
+    public static class Builder {
```

Comment:
I'm concerned about making the ControlsOutput object immutable. In my own bot logic, there are situations where I like to construct output in multiple stages, e.g. one function to determine pitch yaw roll, and another to decide whether to boost. My goal was to make that type of thing ergonomic.

Can you explain more about why you want to go immutable?

Reply:
The main reason why I wanted this particular object to be immutable is that mutable objects, and especially data classes can easily lead to bugs. It is not rare that a single object is handled by several others, and once modified by one, it affects the others which may not want the object to be modified for them. 
Also, based on this example, I thought it was the usual practice to create a new ControlsOutput every frame, thus I did not really see the issue you mentioned. 

Concerning the example you have given, I reckon you could simply pass the builder to your functions instead of an already-created ControlsOutput? When your function has computed what you want, you just call the corresponding withXXX function of the builder. And when you have called all the functions, you eventually build the object.

Basically, if the object gets immutable, I think it will avoid possible bugs such as those I mentioned earlier. On the other hand, it leads to constant instantiation per game tick, which might cause a performance issue. It really is up to you to decide which of the two you prefer. 

If you want to keep ControlsOuput a mutable class, then I'll commit a change to remove the builder.

Reply:
I do believe there is value in keeping it simple. Inexperienced programmers might not fully understand the nuances of immutable classes and factories methods.

Experienced programmers can still choose to modify the example bot to fit their needs, although I am personally not very concerned about the performance hit.

Reply:
Yeah, I agree with kipje on this one, I'm wary of imposing a full builder pattern on novices.

In other situations, e.g. with the vector classes, I'm completely in favor of immutability. In this specific case, people's bot code almost never reads the attributes of ControlsOutput. 99% of the time people will just write and return, never getting into those buggy situations, so I don't think it's worth it.